### PR TITLE
[MNG-7590] Allow to configure resolver by properties in settings.xml

### PR DESCRIPTION
**Pull request by slawekjaranowski**
Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by michael-o** [comment]
Intersting case. Will review next week.

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java [hunk 1]
```diff
@@ -150,6 +152,8 @@ public DefaultRepositorySystemSession newRepositorySession( MavenExecutionReques
         configProps.put( ConfigurationProperties.USER_AGENT, getUserAgent() );
         configProps.put( ConfigurationProperties.INTERACTIVE, request.isInteractiveMode() );
         configProps.put( "maven.startTime", request.getStartTime() );
+        // First add properties populated from settings.xml
+        configProps.putAll( getConfigFromRequestProfiles( request ) );
         // Resolver's ConfigUtils solely rely on config properties, that is why we need to add both here as well.
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:157]: I think this order also complies with preferred inheritance at point 7: https://cwiki.apache.org/confluence/display/MAVEN/Commandline+inheritance

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java [hunk 2]
```diff
@@ -150,6 +152,8 @@ public DefaultRepositorySystemSession newRepositorySession( MavenExecutionReques
         configProps.put( ConfigurationProperties.USER_AGENT, getUserAgent() );
         configProps.put( ConfigurationProperties.INTERACTIVE, request.isInteractiveMode() );
         configProps.put( "maven.startTime", request.getStartTime() );
+        // First add properties populated from settings.xml
+        configProps.putAll( getConfigFromRequestProfiles( request ) );
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:156]: getPropertiesFromRequestedProfiles()?

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java [hunk 3]
```diff
@@ -313,6 +317,28 @@ else if ( !MAVEN_RESOLVER_TRANSPORT_AUTO.equals( transport ) )
...
+            .filter( entry -> isPropertyForResolver( entry.getKey().toString() ) )
+            .collect(
+                Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue, ( k1, k2 ) -> k2  ) );
+    }
+
+    private boolean isPropertyForResolver( String propertyName )
+    {
+        return propertyName != null
+            && ( propertyName.startsWith( "maven.resolver." ) || propertyName.startsWith( "aether." ) );
+    }
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:339]: We don't filter properties from user or system, why here?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java (hunk 3 above)
Comment (slawekjaranowski) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:339]: Good question, @cstamas  what do you think?

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java (hunk 2 above)
Comment (slawekjaranowski) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java]: method renamed

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java (hunk 1 above)
Comment (slawekjaranowski) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:157]: so order is correct

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java (hunk 3 above)
Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:339]: @cstamas WDYT? I think the filter is not required here.

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:339]: unsure about filtering as well... IMO is not needed.

Comment (slawekjaranowski) [reply maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java]: filtering removed

**Review by michael-o — APPROVED** [review]
LGTM. Should also go to 3.9.x, IMHO.

**Comment by michael-o** [comment]
Yet another thing came to my mind: Consider people use custom components in Resolver and they need access to all config props, then filtering would break it. So all is good.

**Comment by cstamas** [comment]
Well, sorry for late as I missed: due https://issues.apache.org/jira/browse/MRESOLVER-284 this is not true for ANY settings.... as we talk here only about session config properties, but if you look at linked issue, some of config properties are expected from PlexusContext/SisuParameters.... this also means that this PR makes it impossible to have these expected as parameters to be set from settings.xml, and this is why I think is better thing to do:
* on short term, to have all these in `$MAVEN_ARGS` or `.mvn/maven.conf` (or mavenrc)
* on long term we need something dedicated I guess

So, again, sorry, but I'd -1 this and rollback the commit as well, but am open to discussion.

**Comment by michael-o** [comment]
> Well, sorry for late as I missed: due https://issues.apache.org/jira/browse/MRESOLVER-284 this is not true for ANY settings.... as we talk here only about session config properties, but if you look at linked issue, some of config properties are expected from PlexusContext/SisuParameters.... this also means that this PR makes it impossible to have these expected as parameters to be set from settings.xml, and this is why I think is better thing to do:
> 
>     * on short term, to have all these in `$MAVEN_ARGS` or `.mvn/maven.conf` (or mavenrc)
> 
>     * on long term we need something dedicated I guess
> 
> 
> So, again, sorry, but I'd -1 this and rollback the commit as well, but am open to discussion.

I don't understand the contradiction. ARen't the settings props read before the session is created? So for those which it can apply it will apply?

**Comment by slawekjaranowski** [comment]
I also think that the most configuration will be worked.
I see that only `aether.syncContext.named.factory` and `aether.syncContext.named.nameMapper`
was moved from session scope to container scope.

But I'm not sure if it was the best step - but it is another topic, I can misunderstood something.

**Comment by cstamas** [comment]
> But I'm not sure if it was the best step - but it is another topic, I can misunderstood something.

Today, sadly after 1.9.0 is out, am also not sure was it the best solution, as it breaks the "session is the config" dogma... and will probably fix it in 1.9.1 or so.... So, basically, this PR may work as is, and I will rework internals of resolver to NOT require this.

**Comment by cstamas** [comment]
I am pulling back my veto, so leave all as is (port it to master probably as well).

**Comment by slawekjaranowski** [comment]
This is on master - here is cherry-pick to 3.9 #867

**Pull request merged**